Inject liborbit.so using UserSpaceInstrumentation lib

pierricgimmig opened the pull request.

danielfenner commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,16 +4,47 @@
+  int32_t pid = capture_options.pid();
+
+  OUTCOME_TRY(AttachAndStopProcess(pid));
+  {
```
The additional scope is not needed here. Restarting on return is just fine.

pierricgimmig replied:
Well that's interesting, if we get rid of the scope, then "result" will always be outcome::success because the scope exit will be executed after the return value has been assigned.

pierricgimmig replied:
Done.

danielfenner commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,16 +4,47 @@
+    OUTCOME_TRY(address, DlsymInTracee(pid, handle, kInitFunction));
+
+    // Initialize all api function tables.
+    for (const ApiTableInfo& info : capture_options.api_table_infos()) {
```
This loop could be moved into the init function, right? Not sure if it is worth it.

pierricgimmig replied:
Oh but we'd have to pass in arrays to the init function, seems complicated, or am I missing something?

danielfenner replied:
Hmm, yes, right. You would need to allocate memory in the tracee for that and pass the pointer. Let's keep it as is. If we really need this at some point we'll come up with some tooling around it.

danielfenner commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,8 +95,38 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
+static void SetupApiTableInfos(const ProcessData& process,
+                               const orbit_client_data::ModuleManager& module_manager,
+                               CaptureOptions* capture_options) {
+  size_t kMaxApiVersion = 0;
```
`constexpr size_t kMaxApiVersion = 0;`

pierricgimmig replied:
Actually replaced by kOrbitApiVersion from Orbit.h.

danielfenner commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,8 +95,38 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
+                               const orbit_client_data::ModuleManager& module_manager,
+                               CaptureOptions* capture_options) {
+  size_t kMaxApiVersion = 0;
+  const char* kOrbitApiPrefix = "g_orbit_api_v";
```
const constexpr char* kOrbitApiPrefix = "g_orbit_api_v";

pierricgimmig replied:
Done.

danielfenner commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,8 +95,38 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
+        if (symbol_info->name() != orbit_api_table_name) continue;
+        ApiTableInfo* api_table_info = capture_options->add_api_table_infos();
+        api_table_info->set_module_path(module_data->file_path());
+        api_table_info->set_api_table_address(symbol_info->address() + base_address.value());
```
Is the load bias missing here?

pierricgimmig replied:
yikes, good point, thanks. I think this is the proper way to use the load bias here?

```
api_table_info->set_api_table_address(symbol_info->address() + base_address.value() -
                                              module_data->load_bias());
```

danielfenner commented on src/UserSpaceInstrumentation/include/UserSpaceInstrumentation/InjectLibraryInTracee.h:
```diff
@@ -5,6 +5,7 @@
 #ifndef USER_SPACE_INSTRUMENTATION_INJECT_LIBRARY_IN_TRACEE_H_
 #define USER_SPACE_INSTRUMENTATION_INJECT_LIBRARY_IN_TRACEE_H_
 
+#include <dlfcn.h>
```
I get it, but a automatic check would complain since this violates iwyu. Suggestion: make `RTLD_NOW` the default for `flag` parameter below.

pierricgimmig replied:
Good idea, thanks, I did just that.

pierricgimmig replied:
Actually I'll submit the change in a separate PR and remove RTLD_NOW from other call sites.

dpallotti replied:
I don't get it, what's the reason for adding the default `RTLD_NOW` value? This to seems to be a case where using a default argument reduces readability at the call sites.

pierricgimmig replied:
The main thing I want to do is include <dlfcn.h> here. Maybe "// IWYU pragma: keep" is what we need?

pierricgimmig replied:
I did just that, although I didn't mind the default parameter.

dimitry- commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,16 +4,47 @@
+  {
+    // Make sure we resume the target process, even on early-outs.
+    orbit_base::unique_resource scope_exit{
+        pid, [&result](int32_t pid) { result = DetachAndContinueProcess(pid); }};
```
I feel like we want a proper scope_guard here (one that does not need a resource)

pierricgimmig replied:
Do we have one?

dimitry- replied:
not yet :)

pierricgimmig replied:
Once we have it I'll be happy to use it :-)

pierricgimmig commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,16 +4,47 @@
+#include "UserSpaceInstrumentation/ExecuteInProcess.h"
+#include "UserSpaceInstrumentation/InjectLibraryInTracee.h"
 
+using namespace orbit_user_space_instrumentation;
```
oh, right, thanks.

pierricgimmig commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,16 +4,47 @@
+
+    // Load liborbit.so and find api table initialization function.
+    const std::string kLibPath = orbit_base::GetExecutableDir() / "liborbit.so";
+    const char* kInitFunction = "orbit_api_initialize";
```
Done.

danielfenner reviewed (approved):
lgtm!

dimitry- commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -79,13 +80,12 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
        max_local_marker_depth_per_command_buffer, collect_memory_info,
        memory_sampling_period_ns]() mutable {
-        return CaptureSync(process_id, module_manager, selected_functions, selected_tracepoints,
+        return CaptureSync(process, module_manager, selected_functions, selected_tracepoints,
```
I would say pass api_table_infos here instead of process.. and generate it before this call, the way we do it for everything else here.

pierricgimmig replied:
Reverted to original signature.

dimitry- commented on src/GrpcProtos/capture.proto:
```diff
@@ -27,6 +27,13 @@ message InstrumentedFunction {
 
+message ApiTableInfo {
+  string module_path = 1;
+  uint64 api_table_address = 2;
```
is this absolute address? maybe we need to pass it in form of file_path and offset, the way it is done for instrumented functions and have the service side to figure out where the module is loaded?

Also this will let us do it on the fly without involvement of the client if a module is not yet loaded.

pierricgimmig replied:
Changed so that we only send the relative address, module_path and build_id. The service verifies the build-id's match and computes the absolute address.

dpallotti commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,15 +4,51 @@
 namespace orbit_api {
+ErrorMessageOr<void> InitializeInTracee(const CaptureOptions& capture_options) {
+  if (capture_options.api_table_infos().size() == 0) {
+    return ErrorMessage("No api table to initilize.");
```
```suggestion
    return ErrorMessage("No api table to initialize.");
```

dpallotti commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,15 +4,51 @@
+
+  // Make sure we resume the target process, even on early-outs.
+  orbit_base::unique_resource scope_exit{pid, [](int32_t pid) {
+                                           if (DetachAndContinueProcess(pid).has_error())
+                                             ERROR("Detaching from %i", pid);
```
Use braces.

pierricgimmig replied:
Done.

pierricgimmig commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,8 +95,38 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
+  for (const ModuleData* module_data : module_manager.GetAllModuleData()) {
+    std::optional<uint64_t> base_address = process.GetModuleBaseAddress(module_data->file_path());
+    if (!base_address.has_value()) {
+      ERROR("No base address found for module %s", module_data->file_path());
```
done.

pierricgimmig commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,8 +95,38 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
+      continue;
+    }
+
+    // TODO: The global variable symbols will be obtained in the next PR.
```
Outdated, added the implementation.

pierricgimmig commented on src/GrpcProtos/capture.proto:
```diff
@@ -27,6 +27,13 @@ message InstrumentedFunction {
   string function_name = 5;
 }
 
+message ApiTableInfo {
```
I changed the name to ApiFunctionInfo which is more self describing. The logic changed as well, we now call a function in the tracee to retrieve the address of the function table.

pierricgimmig commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,8 +95,38 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
   return capture_result;
 }
 
+static void SetupApiTableInfos(const ProcessData& process,
```
Good point, I added a paragraph.

pierricgimmig commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,8 +95,38 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
 
+static void SetupApiTableInfos(const ProcessData& process,
+                               const orbit_client_data::ModuleManager& module_manager,
+                               CaptureOptions* capture_options) {
```
Done.

pierricgimmig commented:
Hi @dpallotti , @dimitry- , @danielfenner . I added a few commits to this PR so that the change works end-to-end with no TODOs. In essence, the changes are:
- addressed comments
- don't rely on global variable symbol, but only on function symbols to find the address where to punch in functions (see FindApiFunctions in CaptureClient.cpp)
- Don't use absolute address when sending ApiFunctionInfos
- Call a function in tracee to get api function table address

Not done: moving api loading logic to ApiLoader project. Will do in separate PR, b/185862867.

I know the PR has gotten a little bigger, but hopefully it is now clearer. Thanks again for your help.

danielfenner commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,15 +4,101 @@
+    if (module_info == nullptr) continue;
+
+    // Get address of function table by calling "orbit_api_get_function_table_address_vN" in tracee.
+    void* api_function_address = absl::bit_cast<void*>(
```
Nothing for this PR but I wonder if `void*` is the right choice to identify addresses in the trace. Both processes are 64 bit so it doesn't matter. I might consider to change the tracee side pointers to uint64_t in InjectLibraryInTracee.h and ExecuteInProcess.h.

pierricgimmig replied:
Yeah it's a good point, we should decide if we want to support 32 bit tracees or not, ideally we would, but then we need to add 32 bit versions for functions in InjectLibraryInTracee.h and ExecuteInProcess.h

dimitry- commented on src/OrbitClientData/ModuleManager.cpp:
```diff
@@ -108,4 +108,12 @@ std::vector<FunctionInfo> ModuleManager::GetOrbitFunctionsOfProcess(
-}  // namespace orbit_client_data
\ No newline at end of file
+std::vector<const ModuleData*> ModuleManager::GetAllModuleData() const {
+  std::vector<const ModuleData*> result;
```
MutexLock

pierricgimmig replied:
Oh yes, thanks! Done.

dimitry- commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,6 +96,34 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
+      std::string orbit_api_function_name = absl::StrFormat("%s%u", kOrbitApiGetAddressPrefix, i);
+      uint64_t function_hash = function_utils::GetHash(orbit_api_function_name);
+
+      const FunctionInfo* function_info = module_data->FindFunctionFromHash(function_hash);
```
maybe add FindFunctionByName (you can use hash internally for now but at least do the verification for cases of hash collision)

pierricgimmig replied:
Yeah good idea, thanks, I did just that.

dimitry- reviewed (approved):
otherwise lgtm

danielfenner commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,15 +4,101 @@
+using orbit_user_space_instrumentation::ExecuteInProcess;
+
+static ErrorMessageOr<absl::flat_hash_map<std::string, ModuleInfo>> GetModulesByPathForPid(
+    int32_t pid) {
```
I use `pid_t` from `#include <sys/types.h>` as a type for `pid`. Probably doesn't matter since it is an int everywhere.

pierricgimmig replied:
yeah, I think it's fine

danielfenner reviewed (approved):
lgtm!

dpallotti commented on src/Service/LinuxTracingHandler.cpp:
```diff
@@ -60,6 +62,13 @@ void LinuxTracingHandler::SetupIntrospection() {
       });
 }
 
+void LinuxTracingHandler::SetupOrbitApi(const CaptureOptions& capture_options) {
```
This should not be in `LinuxTracingHandler` but further up in `OrbitService`, closer to `CaptureServiceImpl::Capture` (probably just in that file, this is really not a lot of code after all).
Basically the idea is that now `LinuxTracingHandler` is not the "master" producer anymore, but is just one of the producers, together with the memory producer and the receivers from the Unix-socket channel (which manual instrumentation also uses).

pierricgimmig replied:
Sounds good, thanks for the thorough review @dpallotti . I moved that logic directly in CaptureServiceImpl::Capture, basically calling 'orbit_api::InitializeInTracee' directly, without the 'SetupOrbitApi' wrapper.

dpallotti commented on src/Service/LinuxTracingHandler.cpp:
```diff
@@ -60,6 +62,13 @@ void LinuxTracingHandler::SetupIntrospection() {
+void LinuxTracingHandler::SetupOrbitApi(const CaptureOptions& capture_options) {
+  auto result = orbit_api::InitializeInTracee(capture_options);
+  if (result.has_error()) {
+    LOG("Warning: Could not initialize Orbit Api: %s", result.error().message());
```
I think this could just be an `ERROR`, we don't have a distinction between `ERROR` and less severe warnings (and this is actually quite severe).

pierricgimmig replied:
Yeah, switched to ERROR, thanks

dpallotti commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,6 +96,34 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
 
+[[nodiscard]] static std::vector<ApiFunction> FindApiFunctions(
+    const orbit_client_data::ModuleManager& module_manager) {
+  // Api functions are declared in Orbit.h. They are implemented in user code through the
```
I would move this comment outside of the block (above the function) for better visibility.

pierricgimmig replied:
Done.

dpallotti commented on src/OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -95,6 +96,34 @@ Future<ErrorMessageOr<CaptureListener::CaptureOutcome>> CaptureClient::Capture(
+    const orbit_client_data::ModuleManager& module_manager) {
+  // Api functions are declared in Orbit.h. They are implemented in user code through the
+  // ORBIT_API_INSTANTIATE macro. Those functions are used to query the tracee for Orbit specific
+  // information such as the memory location where Orbit should write function pointers to enable
```
```suggestion
  // information, such as the memory location where Orbit should write function pointers to enable
```
This is a long sentence.

pierricgimmig replied:
Very good, thanks. Done.

dpallotti commented on src/GrpcProtos/capture.proto:
```diff
@@ -27,6 +27,14 @@ message InstrumentedFunction {
   string function_name = 5;
 }
 
+message ApiFunction {
```
I still feel like a small documentation comment here would help, even if the name changed, because the purpose is quite obscure unless one has reviewed this change. It can also point you to `FindApiFunctions`.

pierricgimmig replied:
Added some doc.

dpallotti commented on src/Api/include/Api/Orbit.h:
```diff
@@ -35,9 +35,9 @@
+// and place the ORBIT_API_INSTANTIATE macro in an implementation file. Orbit will automatically
 // deploy and dynamically load liborbit.so into the target process. Orbit will then write the proper
-// function addresses into the "g_orbit_api_v0" table.
+// function addresses into the "g_orbit_api_v[N]" table.
```
```suggestion
// function addresses into the "g_orbit_api_v<N>" table.
```
Maybe a weird suggestion, but it looked like an array otherwise.

pierricgimmig commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,15 +4,101 @@
+                                         }};
+
+  // Load liborbit.so and find api table initialization function.
+  const std::string kLibPath = orbit_base::GetExecutableDir() / "liborbit.so";
```
Done, thanks.

pierricgimmig commented:
Thanks @danielfenner , @dimitry- and @dpallotti  for the thorough reviews! I addressed all the comments.

dpallotti commented on src/Api/InitializeInTracee.cpp:
```diff
@@ -4,15 +4,101 @@
+  for (const ApiFunction& api_function : capture_options.api_functions()) {
+    // Filter api functions.
+    constexpr const char* kOrbitApiGetAddressPrefix = "orbit_api_get_function_table_address_v";
+    if (absl::StrContains(api_function.name(), kOrbitApiGetAddressPrefix) == false) continue;
```
```suggestion
    if (!absl::StrContains(api_function.name(), kOrbitApiGetAddressPrefix)) continue;
```

dpallotti commented on src/GrpcProtos/capture.proto:
```diff
@@ -27,6 +27,18 @@ message InstrumentedFunction {
 
+// Api functions are declared in Orbit.h. They are implemented in user code
+// through the ORBIT_API_INSTANTIATE macro. Those functions are used to query
+// the tracee for Orbit specific information. See "FindApiFunction" form more
```
```suggestion
// the tracee for Orbit specific information. See "FindApiFunction" for more
```

pierricgimmig replied:
Arrgh, thanks for that.
